File: DependencyAssigment/1.JavaVersion/dependancymangmentassigment/test/test/smartrecruitet/coding/PackageManagerTest.java
package test.smartrecruitet.coding;

import static org.junit.Assert.*;

import java.util.Arrays;
import java.util.List;


import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;
import static org.hamcrest.MatcherAssert.assertThat;

import com.smartrecruiter.coding.*;


public class PackageManagerTest {
	
	private PackageManger classundertest;

	
	@Before
	public void setup(){
		classundertest= new PackageManger();
	}
	
	@Rule
    public ExpectedException thrown = ExpectedException.none();
	
	
	
	@Test
	public void testForCycleDepedecies() {
			
		thrown.expect(RuntimeException.class);
	    thrown.expectMessage("Invalid Dependencies Array : "
			+ "dependency specification must not contains cycles order. ");
		classundertest.getOrderOfDependecy(new String[]{"KittenService: ","Leetmeme: Cyberportal",
                "Cyberportal: Ice","CamelCaser: KittenService",
                "Fraudstream: Leetmeme","Ice: Leetmeme"});
	}
	
	
	@Test
	public void testForParameterForArrayNullReferance() {
			
      assertEquals(null, classundertest.getOrderOfDependecy(null));
	}
	
	@Test
	public void testForParameterForArrayContainNullValue() {
			thrown.expect(NullPointerException.class);
			thrown.expectMessage("Invalid Array : Array must not contain null value");
			classundertest.getOrderOfDependecy(new String[]{"ram: ",null,"hello"});
	}
	
	@Test
	public void testForParameterForEmptyArray() {
		thrown.expect(IllegalArgumentException.class);
		thrown.expectMessage("Invalid Dependencies Array Argument :Format must be Name of a package "
				+ "followed by a colon and space, then any dependencies for example [packagename: dependencyname]");
		classundertest.getOrderOfDependecy(new String[]{"aaa: ","bbb:","", " : "});
	}
	
	
	@Test
	public void testForOrderDependency() {
			
		String testinput[]=new String[]{"KittenService: ","Leetmeme: Cyberportal",
                "Cyberportal: Ice","CamelCaser: KittenService",
                "Fraudstream: Leetmeme","Ice: "};
	    List<String>expexted= Arrays.asList("KittenService", "Ice", "Cyberportal", "Leetmeme", "CamelCaser", "Fraudstream");
       assertEquals(classundertest.getOrderOfDependecy(testinput), expexted);
	
	}
	
	
	@After
	public void cleanup(){
		classundertest=null;
		thrown=null;
		
	}

}

File: DependencyAssigment/2.JavaScriptVersion/test/packagemanagerTest.js


var packagemanager = require('../packagemanager');
var chi = require('chai');
var assert = chi.assert;
var expect = chi.expect;


 /*
 // to see result output remove comment of this section

 var packageArray=['KittenService: ','Leetmeme: Cyberportal',
                  'Cyberportal: Ice','CamelCaser: KittenService',
                  'Fraudstream: Leetmeme','Ice: '];
                   //print ordered  dependency and package   
      console.log(getOrderOfDependecy(packageArray)); 
      */

const testinput1=['KittenService: ','Leetmeme: Cyberportal',
                  'Cyberportal: Ice','CamelCaser: KittenService',
                  'Fraudstream: Leetmeme','Ice: Leetmeme'];
                  
describe('packagemanager', function(){
                                      it('test1 for CycleDepedecies ',function(){
                                      assert.throws(function(){packagemanager.getOrderOfDependecy(testinput1)},
                                       Error,"Invalid Dependencies Array  : "
			                  + "dependency specification must not contains cycles order. ");     
                                                            });
                                                      });
var testinput2;
describe('packagemanager', function(){
                                      it('test For Parameter of function For Null Referance or Undefined ',function(){
                                      assert.equal(packagemanager.getOrderOfDependecy(testinput2),null);     
                                                            });
                                                      });

const testinput3=["ram: ",null,"hello"];

describe('packagemanager', function(){
                                      it('test for parameter for Array Containing NullValue ',function(){
                                      assert.throws(function(){packagemanager.getOrderOfDependecy(testinput3)},
                                       Error,"Invalid Array : Array must not contain null value");     
                                                            });
                                                      });


const testinput4=['KittenService: ','Leetmeme: Cyberportal',
                  'Cyberportal: Ice','CamelCaser: KittenService',
                  'Fraudstream: Leetmeme','Ice: '];
const expectedoutput=['KittenService', 'Ice', 'Cyberportal', 'Leetmeme', 'CamelCaser', 'Fraudstream'];

describe('packagemanager', function(){
                                      it('test for output Order of dependency ',function(){
                  expect(packagemanager.getOrderOfDependecy(testinput4)).to.have.same.members(expectedoutput);   });
                                                      });